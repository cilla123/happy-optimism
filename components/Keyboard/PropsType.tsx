export default interface PropsType {
  type?: 'number' | 'price' | 'idcard';
  onKeyClick?: (key: string) => void;
}
