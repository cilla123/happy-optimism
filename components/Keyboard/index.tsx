import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';
import Icon from '../Icon';

const NUMBER_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '', '0', 'close'];
const PRICE_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '0', 'close'];
const IDCARD_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'x', '0', 'close'];

const stopPropagation = (e) => {
  e.stopPropagation();
  // e.nativeEvent.stopImmediatePropagation();
};

export interface KeyboardProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Keyboard extends PureComponent<KeyboardProps, {}> {

  static defaultProps = {
    prefixCls: 'za-keyboard',
    type: 'number',
  };

  private longPressTimer;

  onLongPressIn = (key) => {
    this.onKeyClick(key);
    this.longPressTimer = setTimeout(() => {
      this.longPressTimer = setInterval(() => {
        this.onKeyClick(key);
      }, 100);
    }, 800);
  }

  onLongPressOut = () => {
    clearInterval(this.longPressTimer);
  }

  onKeyClick = (key) => {
    if (key.length === 0) {
      return;
    }

    const { onKeyClick } = this.props;
    if (typeof onKeyClick === 'function') {
      onKeyClick(key);
    }
  }

  getKeys = () => {
    const { type } = this.props;
    switch (type) {
      case 'price':
        return PRICE_KEYS;

      case 'idcard':
        return IDCARD_KEYS;

      default:
        return NUMBER_KEYS;
    }
  }

  renderKey = (text, index) => {
    const { prefixCls } = this.props;

    const keyCls = classnames(`${prefixCls}-item`, {
      [`${prefixCls}-item-disabled`]: text.length === 0,
    });

    return (
      <div
        className={keyCls}
        key={+index}
        onClick={() => this.onKeyClick(text)}
      >
        {(text === 'close') ? <Icon type="keyboard" /> : text}
      </div>
    );
  }

  render() {
    const { prefixCls } = this.props;

    return (
      <div className={prefixCls} onClick={stopPropagation}>
        <div className={`${prefixCls}-keys`}>
          {this.getKeys().map(this.renderKey)}
        </div>
        <div className={`${prefixCls}-handle`}>
          <div
            className={`${prefixCls}-item`}
            onTouchStart={() => this.onLongPressIn('delete')}
            onTouchEnd={() => this.onLongPressOut()}
          >
            <Icon type="deletekey" />
          </div>
          <div className={`${prefixCls}-item ${prefixCls}-item-ok`} onClick={() => this.onKeyClick('ok')}>确定</div>
        </div>
      </div>
    );
  }
}
