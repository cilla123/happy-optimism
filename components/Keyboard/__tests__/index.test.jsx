import React from 'react';
import { render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Keyboard from '../index';

describe('Keyboard', () => {
  it('renders correctly', () => {
    const wrapper = render(<Keyboard />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is price', () => {
    const wrapper = render(<Keyboard type="price" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is idcard', () => {
    const wrapper = render(<Keyboard type="idcard" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('click keyboard', () => {
    const onKeyClick = jest.fn();
    const wrapper = shallow(<Keyboard onKeyClick={onKeyClick} />);
    const keys = wrapper.find('.za-keyboard-keys');
    expect(keys.childAt(0).text()).toBe('1');
    keys.childAt(0).simulate('click');
    expect(onKeyClick).toBeCalledWith('1');
    keys.childAt(9).simulate('click');
    keys.childAt(11).simulate('click');
    expect(onKeyClick).toBeCalledWith('close');

    const handles = wrapper.find('.za-keyboard-handle');
    handles.childAt(0).simulate('touchstart');
    expect(onKeyClick).toBeCalledWith('delete');
    handles.childAt(1).simulate('click');
    expect(onKeyClick).toBeCalledWith('ok');
  });
});

