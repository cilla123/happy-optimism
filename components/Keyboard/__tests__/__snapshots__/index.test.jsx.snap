// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard renders correctly 1`] = `
<div
  class="za-keyboard"
>
  <div
    class="za-keyboard-keys"
  >
    <div
      class="za-keyboard-item"
    >
      1
    </div>
    <div
      class="za-keyboard-item"
    >
      2
    </div>
    <div
      class="za-keyboard-item"
    >
      3
    </div>
    <div
      class="za-keyboard-item"
    >
      4
    </div>
    <div
      class="za-keyboard-item"
    >
      5
    </div>
    <div
      class="za-keyboard-item"
    >
      6
    </div>
    <div
      class="za-keyboard-item"
    >
      7
    </div>
    <div
      class="za-keyboard-item"
    >
      8
    </div>
    <div
      class="za-keyboard-item"
    >
      9
    </div>
    <div
      class="za-keyboard-item za-keyboard-item-disabled"
    />
    <div
      class="za-keyboard-item"
    >
      0
    </div>
    <div
      class="za-keyboard-item"
    >
      <i
        class="za-icon za-icon-keyboard"
      />
    </div>
  </div>
  <div
    class="za-keyboard-handle"
  >
    <div
      class="za-keyboard-item"
    >
      <i
        class="za-icon za-icon-deletekey"
      />
    </div>
    <div
      class="za-keyboard-item za-keyboard-item-ok"
    >
      确定
    </div>
  </div>
</div>
`;

exports[`Keyboard type is idcard 1`] = `
<div
  class="za-keyboard"
>
  <div
    class="za-keyboard-keys"
  >
    <div
      class="za-keyboard-item"
    >
      1
    </div>
    <div
      class="za-keyboard-item"
    >
      2
    </div>
    <div
      class="za-keyboard-item"
    >
      3
    </div>
    <div
      class="za-keyboard-item"
    >
      4
    </div>
    <div
      class="za-keyboard-item"
    >
      5
    </div>
    <div
      class="za-keyboard-item"
    >
      6
    </div>
    <div
      class="za-keyboard-item"
    >
      7
    </div>
    <div
      class="za-keyboard-item"
    >
      8
    </div>
    <div
      class="za-keyboard-item"
    >
      9
    </div>
    <div
      class="za-keyboard-item"
    >
      x
    </div>
    <div
      class="za-keyboard-item"
    >
      0
    </div>
    <div
      class="za-keyboard-item"
    >
      <i
        class="za-icon za-icon-keyboard"
      />
    </div>
  </div>
  <div
    class="za-keyboard-handle"
  >
    <div
      class="za-keyboard-item"
    >
      <i
        class="za-icon za-icon-deletekey"
      />
    </div>
    <div
      class="za-keyboard-item za-keyboard-item-ok"
    >
      确定
    </div>
  </div>
</div>
`;

exports[`Keyboard type is price 1`] = `
<div
  class="za-keyboard"
>
  <div
    class="za-keyboard-keys"
  >
    <div
      class="za-keyboard-item"
    >
      1
    </div>
    <div
      class="za-keyboard-item"
    >
      2
    </div>
    <div
      class="za-keyboard-item"
    >
      3
    </div>
    <div
      class="za-keyboard-item"
    >
      4
    </div>
    <div
      class="za-keyboard-item"
    >
      5
    </div>
    <div
      class="za-keyboard-item"
    >
      6
    </div>
    <div
      class="za-keyboard-item"
    >
      7
    </div>
    <div
      class="za-keyboard-item"
    >
      8
    </div>
    <div
      class="za-keyboard-item"
    >
      9
    </div>
    <div
      class="za-keyboard-item"
    >
      .
    </div>
    <div
      class="za-keyboard-item"
    >
      0
    </div>
    <div
      class="za-keyboard-item"
    >
      <i
        class="za-icon za-icon-keyboard"
      />
    </div>
  </div>
  <div
    class="za-keyboard-handle"
  >
    <div
      class="za-keyboard-item"
    >
      <i
        class="za-icon za-icon-deletekey"
      />
    </div>
    <div
      class="za-keyboard-item za-keyboard-item-ok"
    >
      确定
    </div>
  </div>
</div>
`;
